Performance analyzer documentation updates (#1318)
opensearch-project/documentation-website: _monitoring-plugins/pa/index.md

---
layout: default
title: Performance analyzer new
parent: pa
nav_order: 58
has_children: true
redirect_from:
  - /monitoring-plugins/pa/
---

# Performance analyzer

Performance analyzer is an agent and REST API that allows you to query numerous performance metrics for your cluster, including aggregations of those metrics. 

The performance analyzer plugin is installed by default in OpenSearch version 2.0 and higher.
{: .note }

## Performance analyzer installation and configuration

Below are the steps to install and configure the performance analyzer plugin.

### Install

> The performance analyzer plugin is included in both the [Docker]({{site.url}}{{site.baseurl}}/opensearch/install/docker/) and [tarball]({{site.url}}{{site.baseurl}}/opensearch/install/tar/) installs of OpenSearch. In the event that you need to install the performance analyzer plugin manually, download the plugin from [Maven](https://search.maven.org/search?q=org.opensearch.plugin). Then install the plugin using the standard [plugins install]({{site.url}}{{site.baseurl}}/opensearch/install/plugins/) process. Performance analyzer will run on each node in a cluster.

Reviewer: Recommended rewrite: 

The performance analyzer plugin is **included in the installation for** [Docker]({{site.url}}{{site.baseurl}}/opensearch/install/docker/) and [tarball]({{site.url}}{{site.baseurl}}/opensearch/install/tar/). **If** you need to install the performance analyzer plugin manually, download the plugin from [Maven](https://search.maven.org/search?q=org.opensearch.plugin) **and install** the plugin using the standard [plugins install]({{site.url}}{{site.baseurl}}/opensearch/install/plugins/) process. Performance analyzer will run on each node in a cluster.

> To start the performance analyzer RCA agent on a tarball install, run the following command.

Reviewer 2: ```suggestion
To start the performance analyzer RCA agent on a tarball install, run the following command:
```

> To start the performance analyzer plugin agent on a tarball install, run the following command.

Reviewer 3: Agent is not plugin and hence we can mention, performance analyzer rca agent

      
````bash
OPENSEARCH_HOME=~/opensearch-2.2.1 OPENSEARCH_JAVA_HOME=~/opensearch-2.2.1/jdk OPENSEARCH_PATH_CONF=~/opensearch-2.2.1/bin ./performance-analyzer-agent-cli
````

> To configure the performance analyzer plugin to auto-start, run the following command.

Reviewer 3: The below command is used to enabled PA plugin and PA RCA agent

> The below command is used to enable the performance analyzer plugin and performance analyzer RCA agent.

Reviewer: Recommended write "The command to enable the performance analyzer plugin and performance analyzer RCA agent is:"

````bash
curl -XPOST localhost:9200/_plugins/_performanceanalyzer/cluster/config -H 'Content-Type: application/json' -d '{"enabled": true}'
````

> To shut down the performance analyzer RCA agent, run the following command.

Reviewer 2: ```suggestion
To shut down the performance analyzer RCA agent, run the following command:
```

> To shutdown the performance analyzer plugin, run the following command.

Reviewer 3: I think you mean to shutdown the PA RCA Agent the below command is used. To disable this plugin the command would be,

```
curl -XPOST localhost:9200/_plugins/_performanceanalyzer/cluster/config -H 'Content-Type: application/json' -d '{"enabled":false}'
```

> To shutdown the performance analyzer RCA agent, run the following command.

Reviewer: Change to "shut down"

````bash
kill $(ps aux | grep -i 'PerformanceAnalyzerApp' | grep -v grep | awk '{print $2}')
````

> To disable the performance analyzer plugin, run the following command.

Reviewer 2: ```suggestion
To disable the performance analyzer plugin, run the following command:
```

````bash

> curl -XPOST localhost:9200/_plugins/_performanceanalyzer/cluster/config -H 'Content-Type: application/json' -d '{"enabled": true}'

Reviewer 3: The value would be false

````

> To uninstall the performance analyzer plugin, run the following command.

Reviewer 2: ```suggestion
To uninstall the performance analyzer plugin, run the following command:
```

````bash
bin/opensearch-plugin remove opensearch-performance-analyzer
````

### Configure performance analyzer

> To configure performance analyzer plugin you will need to edit the `performance-analyzer.properties` configuration file located in the `config/opensearch-performance-analyzer/` directory. Make sure to uncomment the line `#webservice-bind-host` and set it to `0.0.0.0`. You can reference the following example configuration below.

Reviewer: Add "the" after "configure"
Deleted "located" just say "...file in the..."
Delete "below"

"Shard_State":"count","Thread_Waited_Event":"count",
"CB_ConfiguredSize":"B","ThreadPool_QueueCapacity":"count",
"CB_TrippedEvents":"count","Disk_WaitTime":"ms",
"Data_RetryingPendingTasksCount":"count",
"AdmissionControl_CurrentValue":"count",
"Flush_Event":"count","Net_TCP_RxQ":"segments/flow",
"Shard_Size_In_Bytes":"B","Thread_Waited_Time":"s/event",
"HTTP_TotalRequests":"count",
"ThreadPool_ActiveThreads":"count",
"Paging_MinfltRate":"count/s","Net_TCP_SendCWND":"B/flow",
"Cache_Request_Eviction":"count","Segments_Total":"count",
"FollowerCheck_Latency":"ms","Heap_Used":"B",
"Master_ThrottledPendingTasksCount":"count",
"CB_EstimatedSize":"B","Indexing_ThrottleTime":"ms",
"Master_PendingQueueSize":"count",
"Cache_FieldData_Size":"B","Paging_MajfltRate":"count/s",
"ThreadPool_TotalThreads":"count","ShardEvents":"count",
"Net_TCP_NumFlows":"count","Election_Term":"count"}
````

## Root cause analysis

The [root cause analysis]({{site.url}}{{site.baseurl}}/monitoring-plugins/pa/rca/index/) (RCA) framework uses the information from performance analyzer to alert administrators about the root cause of performance and availability issues that their clusters might be experiencing.

> ### Enable

Reviewer 4: ```suggestion
### Enable the RCA framework
```

> To enable the root cause analyzer (RCA) framework, run the following command.

Reviewer 4: ```suggestion
To enable the RCA framework, run the following command:
```

```bash
curl -XPOST http://localhost:9200/_plugins/_performanceanalyzer/rca/cluster/config -H 'Content-Type: application/json' -d '{"enabled": true}'
```

> If you run into the `curl: (52) Empty reply from server` response, run the command below to enable RCA.

Reviewer 4: ```suggestion
If you run into the `curl: (52) Empty reply from server` response, run the following command to enable RCA:
```

> If you run into the `curl: (52) Empty reply from server` response, run the command below to enable RCA.

Reviewer 4: Change "run into" to either "receive" or "encounter".

```bash
curl -XPOST https://localhost:9200/_plugins/_performanceanalyzer/rca/cluster/config -H 'Content-Type: application/json' -d '{"enabled": true}' -u 'admin:admin' -k
```

### Example API query and response

> Request all available RCAs.

Reviewer 4: Let's make this a complete sentence. How about "To request all available RCAs, run the following command:"?

````bash
GET localhost:9600/_plugins/_performanceanalyzer/rca
````

> Request a specific RCA.

Reviewer 4: Make a complete sentence. "To request a specific RCA, run the following command:".

````bash
GET localhost:9600/_plugins/_performanceanalyzer/rca?name=HighHeapUsageClusterRCA
````

> Example response.

Reviewer 4: ```suggestion
The following is an example response:
```

```json
{
  "HighHeapUsageClusterRCA": [{
    "RCA_name": "HighHeapUsageClusterRCA",
    "state": "unhealthy",
    "timestamp": 1587426650942,
    "HotClusterSummary": [{
      "number_of_nodes": 2,
      "number_of_unhealthy_nodes": 1,
      "HotNodeSummary": [{
        "host_address": "192.168.144.2",
        "node_id": "JtlEoRowSI6iNpzpjlbp_Q",
        "HotResourceSummary": [{
          "resource_type": "old gen",
          "threshold": 0.65,
          "value": 0.81827232588145373,
          "avg": NaN,
          "max": NaN,
          "min": NaN,
          "unit_type": "heap usage in percentage",
          "time_period_seconds": 600,
          "TopConsumerSummary": [{
              "name": "CACHE_FIELDDATA_SIZE",
              "value": 590702564
            },
            {
              "name": "CACHE_REQUEST_SIZE",
              "value": 28375
            },
            {
              "name": "CACHE_QUERY_SIZE",
              "value": 12687
            }
          ],
        }]
      }]
    }]
  }]
}
```

> ## Performance analyzer and root cause analysis API references

Reviewer 4: ```suggestion
## Performance analyzer and RCA API references
```

### Related links

> Further documentation on the use of performance analyzer and root cause analysis can be found at the following links.

Reviewer 4: ```suggestion
Further documentation on the use of performance analyzer and RCA can be found at the following links:
```

GET localhost:9600/_plugins/_performanceanalyzer/RCA?name=HighHeapUsageClusterRCA
````

Example response.

```json
{
  "HighHeapUsageClusterRCA": [{
    "RCA_name": "HighHeapUsageClusterRCA",
    "state": "unhealthy",
    "timestamp": 1587426650942,
    "HotClusterSummary": [{
      "number_of_nodes": 2,
      "number_of_unhealthy_nodes": 1,
      "HotNodeSummary": [{
        "host_address": "192.168.144.2",
        "node_id": "JtlEoRowSI6iNpzpjlbp_Q",
        "HotResourceSummary": [{
          "resource_type": "old gen",
          "threshold": 0.65,
          "value": 0.81827232588145373,
          "avg": NaN,
          "max": NaN,
          "min": NaN,
          "unit_type": "heap usage in percentage",
          "time_period_seconds": 600,
          "TopConsumerSummary": [{
              "name": "CACHE_FIELDDATA_SIZE",
              "value": 590702564
            },
            {
              "name": "CACHE_REQUEST_SIZE",
              "value": 28375
            },
            {
              "name": "CACHE_QUERY_SIZE",
              "value": 12687
            }
          ],
        }]
      }]
    }]
  }]
}
```

## Performance analyzer and root cause analysis API references

> Further documentation on the use of performance analyzer and root cause analysis can be found at the following links.

Reviewer: Add heading "Related links"

        "HotResourceSummary": [{
          "resource_type": "old gen",
          "threshold": 0.65,
          "value": 0.81827232588145373,
          "avg": NaN,
          "max": NaN,
          "min": NaN,
          "unit_type": "heap usage in percentage",
          "time_period_seconds": 600,
          "TopConsumerSummary": [{
              "name": "CACHE_FIELDDATA_SIZE",
              "value": 590702564
            },
            {
              "name": "CACHE_REQUEST_SIZE",
              "value": 28375
            },
            {
              "name": "CACHE_QUERY_SIZE",
              "value": 12687
            }
          ],
        }]
      }]
    }]
  }]
}
```

## Performance analyzer and root cause analysis API references

### Related links

Further documentation on the use of performance analyzer and root cause analysis can be found at the following links.

- [Performance analyzer API guide]({{site.url}}{{site.baseurl}}/monitoring-plugins/pa/api/).
- [Root cause analysis]({{site.url}}{{site.baseurl}}/monitoring-plugins/pa/rca/index/).
- [Root cause analysis API guide]({{site.url}}{{site.baseurl}}/latest/monitoring-plugins/pa/rca/api/).

> - [RFC: Root cause analysis](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/docs/rfc-RCA.pdf).

Reviewer 2: ```suggestion
- [RFC: Root cause analysis](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/docs/rfc-rca.pdf).
```
